### PR TITLE
Surface feasibility in solve results (PrimalStatus, source objective value, auto-report)

### LINKED ISSUE #223: Surface feasibility in solve results (PrimalStatus, source objective value, auto-report)
## Motivation
MQT QAO (arXiv:2406.12840) auto-runs `check_constraint_optimal_solution` after every solve, and its Solution object exposes `check_constraint_all_solutions` and a valid-solution rate (`src/mqt/qao/solvers.py:1503-1530`; `docs/Solution.rst:27-41`) — albeit with operator-precedence bugs in its AND/OR checkers (`constraints.py:402,458,557,613`). Our checking machinery is sounder (numeric `MOIU.distance_to_set` residuals, `feasibility.jl:320-331`) but is not wired into the solve path, so practitioners get misleading statuses by default.

## Current state
Machinery exists: `violations`/`is_feasible`/`feasibility_report` (`src/analysis/feasibility.jl:69-117`) — and note the aggregate valid-sample rate ALREADY exists: `feasibility_report(model)` with `result=nothing` summarizes every result with `feasible_count`/`result_count` plus violation histograms (`feasibility.jl:48-54, 102-117, 255-277`). The confirmed gaps: (1) nothing runs automatically after `optimize!` (`src/wrapper.jl:27-44`); (2) `PrimalStatus` is unconditionally `FEASIBLE_POINT` (QUBODrivers `src/library/sampler/wrappers/moi.jl:29-39`; ToQUBO `src/attributes/solver.jl:124-138` just delegates); (3) `MOI.ObjectiveValue` returns the penalized QUBO energy, not the source objective (`src/attributes/solver.jl:150-156`).

## Proposed change
- Override `MOI.get(::Virtual.Model, ::MOI.PrimalStatus)` in `src/attributes/solver.jl` to consult the feasibility analysis for the given result index (cache per result; guard on `ResultCount > 0`).
- Add penalty-free source-objective reporting: evaluate `model.f` on the sample state (same mechanism as `_evaluate_pbf` in `src/reformulation.jl`) — either override `MOI.ObjectiveValue` (behind an attribute, to avoid silently changing semantics) or add a dedicated `Attributes.SourceObjectiveValue`.
- Optionally auto-run and store a `FeasibilityReport` after `optimize!` in `src/wrapper.jl:27-44`, gated by an attribute, so JuMP users get the valid rate without extra calls (`ext/ToQUBOJuMPExt.jl:13-24` wrappers already exist).

## Acceptance criteria
- [ ] Infeasible samples report `INFEASIBLE_POINT` (or documented equivalent); feasible ones unchanged.
- [ ] Source-objective value queryable per result and free of penalty terms (tested against manual evaluation).
- [ ] Aggregate valid rate reachable in one call post-solve; documented in the manual.
- [ ] No behavior change when the gating attributes are off.

Part of the MQT QAO parity epic: https://github.com/JuliaQUBO/QUBO.jl/issues/69

### PR DESCRIPTION
**Pull request by bernalde**
Closes #223

Part of the MQT QAO parity epic: JuliaQUBO/QUBO.jl#69

## Summary

Samplers solve the unconstrained target model, so QUBODrivers unconditionally reports `FEASIBLE_POINT` for every in-range result (`src/library/sampler/wrappers/moi.jl:30-40`), and `MOI.ObjectiveValue` reports the penalized target energy. This PR wires ToQUBO's existing feasibility machinery (#138) into the solve path so results surface source-model semantics:

- **`MOI.PrimalStatus` feasibility check** — new `Attributes.PrimalFeasibilityCheck()` (default **on**): when the sampler reports `FEASIBLE_POINT`, ToQUBO re-checks the decoded sample against source constraints (lazily, cached per result index) and reports `INFEASIBLE_POINT` on violation. Turning the attribute off restores the previous behavior exactly.
- **Penalty-free source objective** — new `Attributes.SourceObjectiveValue(result_index)` plus exported `source_objective_value(model; result)`: evaluates the *source* objective function on the projected source-variable values. `MOI.ObjectiveValue` semantics are deliberately unchanged (the issue offered either option; silently changing a core MOI attribute's meaning seemed riskier than a dedicated, discoverable attribute).
- **Automatic feasibility report** — new `Attributes.AutoFeasibilityReport()` (default off): after each solve, computes the full `FeasibilityReport`, caches it, and warns (respecting `Attributes.Warnings()`) when any sample violates a source constraint. Default-argument `feasibility_report(model)` calls now return the per-solve cached report; caches are invalidated in `MOI.optimize!` and `Compiler.reset!`.

## Design notes

- Default-**on** for `PrimalFeasibilityCheck` is a deliberate behavior change: reporting `FEASIBLE_POINT` for constraint-violating samples is the misleading default the issue targets. Existing suite assertions only check result 1 of well-penalized models and are unaffected.
- The feasibility tolerance is the analysis default (`1e-6`), now shared via `_FEASIBILITY_ATOL` so `violations`/`is_feasible`/`feasibility_report` and the status check agree.
- `MOI.DualStatus` handling is untouched (split out of the former `Union` method).

## Acceptance criteria (from #223)

- [x] Infeasible samples report `INFEASIBLE_POINT`; feasible ones unchanged (`test_feasibility_primal_status`)
- [x] Source-objective value queryable per result and free of penalty terms, tested against manual evaluation (`test_feasibility_source_objective_value`)
- [x] Aggregate valid rate reachable in one call post-solve; documented in the manual (`docs/src/manual/3-results.md`, "Feasibility in Solve Results")
- [x] No behavior change when the gating attributes are off

## Testing

- `julia --project -e 'using Pkg; Pkg.test()'` (Julia 1.10.11): full suite, including new `test_feasibility_primal_status`, `test_feasibility_source_objective_value`, `test_feasibility_auto_report`, and attribute round-trip/reset-survival tests.
- Test-only dependency added: `Logging` (stdlib) for `@test_logs min_level` assertions.

## Branch Hygiene

- Base: `main` (default), branched from `origin/main` at `c5191cb`.
- Open sibling PRs #229 and #230 are Dependabot bumps touching `.github/workflows/` files only — no overlap with this PR's source, tests, or docs. The two historical feasibility branches (#138, #46) are merged and this builds on their machinery.

### CODE REVIEW

**Review by bernalde — COMMENTED** [review]
Maintainer review of head `ecf3854` (author account, so this is a COMMENT review; it cannot serve as a formal approval).

## Scope and design

The PR resolves #223's three acceptance criteria with an opt-out `PrimalStatus` feasibility check, a dedicated penalty-free `SourceObjectiveValue` attribute, and an opt-in post-solve report. Design choices reviewed and accepted:

- Default-**on** `PrimalFeasibilityCheck` is a deliberate behavior change: QUBODrivers hardcodes `FEASIBLE_POINT` for every in-range result, which is misleading for penalized reformulations. The escape hatch restores the raw status exactly, and no existing test asserted the old behavior for an infeasible sample.
- Keeping `MOI.ObjectiveValue` semantics untouched (dedicated attribute instead) avoids silently changing the meaning of a core MOI attribute mid-0.6.x.
- Lazy per-result feasibility caching bounds the `PrimalStatus` overhead to one constraint sweep per queried result per solve; caches are invalidated in both `MOI.optimize!` and `Compiler.reset!`.

## Findings

**Nonblocking: `Compiler.reset!` cache clearing lacks a direct test** 
`reset!` now deletes `:feasibility_report`/`:primal_feasibility` from `moi_settings` (`src/compiler/compiler.jl`), but only the `optimize!`-path invalidation is asserted (the re-optimize identity checks in `test/unit/feasibility.jl`). A regression that drops the `reset!` deletions would go unnoticed until someone hits a stale report after `MOI.empty!`. Concrete fix: in the attributes tests, seed `model.moi_settings[:feasibility_report]` on a fresh `ToQUBO.Optimizer{Float64}()`, call `ToQUBO.Compiler.reset!(model)`, and assert the key is gone.

**Nonblocking: result-indexed attribute should raise `MOI.ResultIndexBoundsError`** 
`MOI.get(model, Attributes.SourceObjectiveValue(99))` raises a plain `ErrorException` via `_check_result_index`, while MOI's own result attributes (`MOI.ObjectiveValue`, `MOI.VariablePrimal`) raise `MOI.ResultIndexBoundsError` — callers catch by type, and JuMP users will see `objective_value(model; result = 99)` and `source_objective_value(model; result = 99)` fail differently. Trade-off: the sibling analysis helpers (`violations`, `is_feasible`) use plain errors, so full consistency is impossible either way. Recommendation: throw `MOI.ResultIndexBoundsError(attr, n)` in the `MOI.get` method for the attribute path (the MOI-conventional behavior), and leave the kwarg-based analysis helpers as they are. Update the `@test_throws` in `test_feasibility_source_objective_value` accordingly.

**Nonblocking: new attributes not discoverable from the settings manual** 
`docs/src/manual/4-settings.md` is where compiler attributes are cataloged, but `PrimalFeasibilityCheck`/`AutoFeasibilityReport` are documented only in `3-results.md`. Add a short cross-reference (a sentence or a two-row pointer under the ToQUBO-specific settings section) so users browsing settings discover the solve-result attributes.

## Verification performed

- Full local suite at exactly this tree (Julia 1.10.11, `Pkg.test()`): 1822 pass, 0 fail.
- CI green on head `ecf3854`: Julia 1.10/1 × ubuntu/windows, docs build + preview deploy (executes the new `@example` blocks in `3-results.md`), labeler.
- Standalone repro script confirming: check-on downgrade to `INFEASIBLE_POINT`, raw `FEASIBLE_POINT` via backend after opt-out, deterministic round-trip after re-optimize; the earlier `NO_SOLUTION` observation is JuMP's `is_model_dirty` bookkeeping, correctly handled in tests/docs by re-optimizing after `set_attribute`.
- File-set match between local merge-base diff and `gh pr diff`; base `main` unchanged since branch point, so PR-head checks equal merge-result checks. No open sibling PRs touch these files.

## Summary

- Blocking: none.
- Nonblocking: 3 (direct `reset!` cache test, `ResultIndexBoundsError` for the attribute path, settings-manual cross-reference).
- Questions: none.
- Merge-readiness: CI green, `mergeStateStatus: CLEAN`, linked issue fully resolved (`Closes #223`). This COMMENT review cannot formally approve; if branch protection requires an approving review, another eligible account must provide it.

**Comment by bernalde** [comment]
Addressed all three nonblocking findings from the [maintainer review](https://github.com/JuliaQUBO/ToQUBO.jl/pull/228#pullrequestreview-4835616737) in fc9f230:

- `reset-cache-test` — added a "Feasibility cache reset" testset in `test/unit/attributes/attributes.jl` that seeds `moi_settings[:feasibility_report]`/`[:primal_feasibility]` on a fresh optimizer, calls `ToQUBO.Compiler.reset!`, and asserts both keys are deleted.
- `result-index-error-type` — `MOI.get(::Virtual.Model, ::Attributes.SourceObjectiveValue)` now calls `MOI.check_result_index_bounds`, so out-of-range result indices raise `MOI.ResultIndexBoundsError` exactly like `MOI.ObjectiveValue`/`MOI.VariablePrimal`; the kwarg-based analysis helpers keep their existing plain errors, as recommended. `test_feasibility_source_objective_value` now asserts the MOI error type.
- `settings-docs-xref` — added a "Solve-Result Feasibility" section to `docs/src/manual/4-settings.md` rendering the `PrimalFeasibilityCheck`/`AutoFeasibilityReport`/`SourceObjectiveValue` docstrings via `@docs`, with a cross-reference to the usage examples in the results manual.

A verification pass over fc9f230 additionally found one PR-introduced docs regression: the exported `source_objective_value` docstring was not rendered in any `@docs` block (the base build had zero missing-docstring warnings). Fixed in fc2d58b by adding it to the results-manual `@docs` list.

Tests: full local suite (`Pkg.test()`, Julia 1.10.11) green at fc2d58b (1820 tests), zero failures. The reset-cache test was mutation-verified: removing the `reset!` cache deletions fails exactly that testset while the other 1818 tests stay green.

No findings were declined.

**Review by bernalde — COMMENTED** [review]
Re-review of head `fc2d58b` after the two follow-up commits. This account authored the PR, so GitHub accepts this as a COMMENT review rather than a formal approval or request-changes verdict.

## Prior findings

All three findings from the `ecf3854` review are resolved at this head:

- `reset-cache-test`: a direct `Compiler.reset!` cache test was added; removing the two cache deletions now fails both targeted assertions.
- `result-index-error-type`: `SourceObjectiveValue` now uses `MOI.check_result_index_bounds`; restoring the old check makes the updated test fail with `ErrorException` instead of `MOI.ResultIndexBoundsError`.
- `settings-docs-xref`: the settings manual now renders all three new attribute docstrings and links to the results examples; the exported `source_objective_value` docstring is also rendered.

## Findings

One Blocking finding is posted inline on `test_feasibility_auto_report`.

**Nonblocking: refresh the sibling-PR statement in the description** 

The Branch Hygiene section says there are no open sibling PRs, but #229 and #230 are now open. They only modify workflow files and do not overlap this PR, so there is no code-coordination risk; update the sentence to name them as non-overlapping workflow-only changes, or remove the absolute claim so the merged record is accurate.

## Verification performed

- Local merge-base and GitHub file sets match exactly: 13 changed files. Current `main` (`c5191cb`) is the merge base, so PR-head checks also exercise the merge-result tree.
- Full local suite at `fc2d58b`, Julia 1.10.11: `Pkg.test()` passed 1820/1820 tests.
- Reviewer-authored targeted probe passed 8/8 checks covering infeasible/opt-out `PrimalStatus`, penalty-free source objective, result-index error type, cache reset, and a zero-objective edge case.
- Mutation checks confirmed the new reset-cache and result-index assertions fail when their fixes are removed. A separate mutation exposed the Blocking disabled-auto-report coverage gap described inline.
- Exact-head GitHub checks are green: all four required Julia 1.10/1 × Ubuntu/Windows jobs, Documentation build, labeler, and `documenter/deploy`.
- Open PRs #229 and #230 touch workflow files only; neither overlaps this PR's source, tests, or documentation.

## Summary

- Blocking: 1 (disabled `AutoFeasibilityReport` behavior is not regression-tested).
- Nonblocking: 1 (stale sibling-PR claim in the description).
- Questions: none.
- Merge-readiness: implementation and current checks are green, but the explicit opt-out acceptance criterion needs the regression assertion above. `main` requires zero approving reviews, while required conversation resolution means the inline thread must be addressed and resolved.

I would not merge this until the blocking issue above is addressed.

File: test/unit/feasibility.jl [hunk 1]
```diff
@@ -219,12 +234,114 @@ function test_feasibility_error_paths()
...
+
+    return nothing
+end
+
+function test_feasibility_auto_report()
+    model, _ = _under_penalized_model()
+
+    @test get_attribute(model, Attributes.AutoFeasibilityReport()) === false
+
+    optimize!(model)
```
Comment (bernalde) [inline test/unit/feasibility.jl]: Blocking: Add a regression assertion for the disabled auto-report path.

The test confirms that the attribute defaults to `false`, but after this `optimize!` it immediately calls `feasibility_report` itself, so it never proves that the solve avoided auto-computing/caching a report or emitting a warning. I removed the early `AutoFeasibilityReport` guard in `_auto_feasibility_report!`; this test still passed 8/8, while the disabled solve emitted `2 of 8 sampled result(s) violate source constraints` and populated the cache. That leaves the acceptance criterion “No behavior change when the gating attributes are off” unguarded.

Please make this first solve assert no warning (for example, `@test_logs min_level = Logging.Warn optimize!(model)`) and, before the explicit `feasibility_report` call, assert that `JuMP.unsafe_backend(model).moi_settings` has no `:feasibility_report` entry. Then the mutation above should fail this test.

**Comment by bernalde** [comment]
Addressed the round-2 review (head `fc2d58b`) in 7c60e67:

- **Blocking — disabled auto-report path unguarded** (inline on `test_feasibility_auto_report`): implemented exactly as requested. The first solve is now wrapped in `@test_logs min_level = Logging.Warn optimize!(model)`, and before the explicit `feasibility_report` call the test asserts `JuMP.unsafe_backend(model).moi_settings` has no `:feasibility_report` entry. Mutation-verified: removing the `AutoFeasibilityReport` guard in `_auto_feasibility_report!` now fails both new assertions (log test + cache-key test) while the other 1820 tests stay green; with the guard intact the full suite passes.
- **Nonblocking — stale sibling-PR claim** (`stale-sibling-pr-claim`): the Branch Hygiene bullet now names #229/#230 as Dependabot bumps touching `.github/workflows/` files only, with no overlap with this PR's source, tests, or docs (verified against their live changed-file lists).

Tests: full local suite (`Pkg.test()`, Julia 1.10.11) green at 7c60e67; CI on the new head is running.

No findings were declined. Thread resolution is left to the human reviewer; `mergeStateStatus` will remain `BLOCKED` until the inline conversation is resolved.

File: test/unit/feasibility.jl (hunk 1 above)
Comment (bernalde) [reply test/unit/feasibility.jl]: Addressed in 7c60e67, exactly as suggested: the first (default-off) solve is now `@test_logs min_level = Logging.Warn optimize!(model)`, and the test asserts `!haskey(JuMP.unsafe_backend(model).moi_settings, :feasibility_report)` before the explicit `feasibility_report` call. Re-ran your mutation — removing the `AutoFeasibilityReport` guard in `_auto_feasibility_report!` now fails both new assertions (1820 other tests green); full suite green with the guard intact.
